File: types/policy.ts
/**
 * Author: Deepankar Das
 */

/**
 * Enforcer — Policy Types
 *
 * Defines the PolicyDecision contract returned by the daemon to enforcement
 * points, and the PolicyRule schema used in YAML policy bundles.
 *
 * TDD Reference: Appendix B.4 (PolicyDecision), Section 8
 * PRD Reference: Appendix C, R-2
 */

import { z } from "zod";

// ── Decision Types ──────────────────────────────────────────────────────────

export const PolicyDecisionType = z.enum([
  "allow",
  "deny",
  "require_approval",
  "allow_degraded",    // Fail-open mode when policy cache is unavailable
  "redact",            // Phase 2
  "quarantine",        // Phase 2
  "simulate",          // Phase 2 (dry-run / log-only)
]);
export type PolicyDecisionType = z.infer<typeof PolicyDecisionType>;

// ── Policy Scope Level ──────────────────────────────────────────────────────

export const PolicyScopeLevel = z.enum([
  "organization",
  "team",
  "repository",
  "local",
]);
export type PolicyScopeLevel = z.infer<typeof PolicyScopeLevel>;

// ── Policy Effect (from YAML rule) ──────────────────────────────────────────

export const PolicyEffectSchema = z.object({
  decision: PolicyDecisionType,
  reason_code: z.string().min(1),
  reason_human: z.string().min(1),
});
export type PolicyEffect = z.infer<typeof PolicyEffectSchema>;

// ── Policy Rule (from YAML bundle) ──────────────────────────────────────────

export const PolicyRuleSchema = z.object({
  policy_id: z.string().min(1),
  version: z.string().min(1),
  scope: z.object({
    level: PolicyScopeLevel,
  }),
  subject: z.object({
    agent_types: z.array(z.string()).default(["*"]),
    users: z.array(z.string()).default(["*"]),
  }),
  action: z.object({
    types: z.array(z.string()),
  }),
  resource: z.record(z.unknown()).default({}),
  conditions: z.record(z.unknown()).default({}),
  effect: PolicyEffectSchema,
  logging: z.object({
    mode: z.enum(["full", "metadata_only", "redacted"]).default("full"),
  }).default({ mode: "full" }),
  approval: z.object({
    required: z.boolean().default(false),
  }).default({ required: false }),
});
export type PolicyRule = z.infer<typeof PolicyRuleSchema>;

// ── Policy Bundle (collection of rules with version) ────────────────────────

export const PolicyBundleSchema = z.object({
  bundle_version: z.string().min(1),
  scope_level: PolicyScopeLevel,
  rules: z.array(PolicyRuleSchema),
});
export type PolicyBundle = z.infer<typeof PolicyBundleSchema>;

// ── PolicyDecision (returned to enforcement point) ──────────────────────────

export const PolicyDecisionSchema = z.object({
  request_id: z.string().min(1),
  decision: PolicyDecisionType,
  reason_code: z.string().min(1),
  reason_human: z.string().min(1),
  policy_id: z.string().min(1),
  policy_version: z.string().min(1),
  approval_required: z.boolean(),
});
export type PolicyDecision = z.infer<typeof PolicyDecisionSchema>;

File: go/internal/types/policy.go
/**
 * Author: Deepankar Das
 */

package types

// PolicyDecisionType enumerates possible policy decisions.
type PolicyDecisionType string

const (
	DecisionAllow           PolicyDecisionType = "allow"
	DecisionDeny            PolicyDecisionType = "deny"
	DecisionRequireApproval PolicyDecisionType = "require_approval"
	DecisionAllowDegraded   PolicyDecisionType = "allow_degraded"
	DecisionRedact          PolicyDecisionType = "redact"
	DecisionQuarantine      PolicyDecisionType = "quarantine"
	DecisionSimulate        PolicyDecisionType = "simulate"
)

// PolicyScopeLevel defines the hierarchy level of a policy rule.
type PolicyScopeLevel string

const (
	ScopeOrganization PolicyScopeLevel = "organization"
	ScopeTeam         PolicyScopeLevel = "team"
	ScopeRepository   PolicyScopeLevel = "repository"
	ScopeLocal        PolicyScopeLevel = "local"
)

// PolicyEffect specifies the enforcement action for a matched rule.
type PolicyEffect struct {
	Decision    PolicyDecisionType `json:"decision" yaml:"decision"`
	ReasonCode  string             `json:"reason_code" yaml:"reason_code"`
	ReasonHuman string             `json:"reason_human" yaml:"reason_human"`
}

// PolicyRule defines a single governance rule.
type PolicyRule struct {
	PolicyID string `json:"policy_id" yaml:"policy_id"`
	Version  string `json:"version" yaml:"version"`
	Scope    struct {
		Level PolicyScopeLevel `json:"level" yaml:"level"`
	} `json:"scope" yaml:"scope"`
	Subject struct {
		AgentTypes []string `json:"agent_types" yaml:"agent_types"`
		Users      []string `json:"users" yaml:"users"`
	} `json:"subject" yaml:"subject"`
	Action struct {
		Types []string `json:"types" yaml:"types"`
	} `json:"action" yaml:"action"`
	Resource   map[string]interface{} `json:"resource" yaml:"resource"`
	Conditions map[string]interface{} `json:"conditions" yaml:"conditions"`
	Effect     PolicyEffect           `json:"effect" yaml:"effect"`
	Logging    struct {
		Mode string `json:"mode" yaml:"mode"`
	} `json:"logging" yaml:"logging"`
	Approval struct {
		Required bool `json:"required" yaml:"required"`
	} `json:"approval" yaml:"approval"`
	Enabled *bool `json:"enabled,omitempty" yaml:"enabled,omitempty"`
}

// IsEnabled returns whether the rule is active (defaults to true if not set).
func (r *PolicyRule) IsEnabled() bool {
	if r.Enabled == nil {
		return true
	}
	return *r.Enabled
}

// PolicyBundle is a versioned collection of policy rules.
type PolicyBundle struct {
	BundleVersion string           `json:"bundle_version" yaml:"bundle_version"`
	ScopeLevel    PolicyScopeLevel `json:"scope_level" yaml:"scope_level"`
	Rules         []PolicyRule     `json:"rules" yaml:"rules"`
}

// PolicyDecision is the daemon's response to an enforcement point.
type PolicyDecision struct {
	RequestID        string             `json:"request_id"`
	Decision         PolicyDecisionType `json:"decision"`
	ReasonCode       string             `json:"reason_code"`
	ReasonHuman      string             `json:"reason_human"`
	PolicyID         string             `json:"policy_id"`
	PolicyVersion    string             `json:"policy_version"`
	ApprovalRequired bool               `json:"approval_required"`
}
